### PR TITLE
[Cocoa] When accessibility bold is enabled, font-family:system-ui is supposed to be bold

### PR DESCRIPTION
**Pull request by litherum**
#### 3e3f9ab0f79b58eb597037a6de8ebf192ddc1728
<pre>
[Cocoa] When accessibility bold is enabled, font-family:system-ui is supposed to be bold
<a href="https://bugs.webkit.org/show_bug.cgi?id=237817">https://bugs.webkit.org/show_bug.cgi?id=237817</a>
&lt;rdar://88249241&gt;

Reviewed by Simon Fraser.

This re-implements the accessibility bold setting in web content. The main logic here is fairly
straightforward:
1. When determining the weight a font should have, use _AXSEnhanceTextLegibilityEnabled() to
       determine the accessibility bold setting is enabled
2. If it is, round-trip the weight we&apos;re setting through CTFontGetAccessibilityBoldWeightOfWeight()
3. Because CTFontGetAccessibilityBoldWeightOfWeight() operates on Core Text font weights, rather
       than CSS weights, wrap that call with denormalizeCTWeight() and normalizeCTWeight().

Because the output of _AXSEnhanceTextLegibilityEnabled() only changes when a notification is fired,
we don&apos;t actually need to call it every time we set the weight of a font. Instead, we can just call
it once and cache its output in a NeverDestroyed&lt;bool&gt;. This is what
platformShouldEnhanceTextLegibility() does. When the notification is fired, we just update the
NeverDestroyed to hold the new value.

We also need a way to mock the accessibility bold setting, for testing. This is done via another
NeverDestroyed, which holds a tri-state: either the accessibility setting was not overridden by
tests, or if it was, it holds the boolean value that it was overriden with. This lives in
overrideEnhanceTextLegibilityStorage().

WebKit Legacy doesn&apos;t use the override mechanism, because there&apos;s no sandbox, so this code can
reliably expect the AX notification to be delivered and for _AXSEnhanceTextLegibilityEnabled() to
work. However, WebKit2 has to plumb this setting through the UI process. When the web process
receives this value from the UI process, it uses the override mechanism to set the value for the
web process. This design is pretty common in WebKit2.

Tests: fast/text/accessibility-bold-system-font/accessibility-bold-system-font.html
       fast/text/accessibility-bold-system-font/accessibility-bold-system-font-2.html
       fast/text/accessibility-bold-system-font/accessibility-bold-system-font-3.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/text/accessibility-bold-system-font-2-expected.html:
* LayoutTests/fast/text/accessibility-bold-system-font-3-expected-mismatch.html: Added.
* LayoutTests/fast/text/accessibility-bold-system-font-3.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::fontIsSystemFont):
(WebCore::overrideEnhanceTextLegibilityStorage):
(WebCore::overrideEnhanceTextLegibility):
(WebCore::platformShouldEnhanceTextLegibility):
(WebCore::shouldEnhanceTextLegibility):
(WebCore::preparePlatformFont):
(WebCore::FontCache::platformInit):
(WebCore::FontCache::platformInvalidate):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebCore/testing/InternalSettings.cpp:
(WebCore::InternalSettings::setShouldMockBoldSystemFontForAccessibility):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/Shared/AccessibilityPreferences.cpp:
(IPC::ArgumentCoder&lt;WebKit::AccessibilityPreferences&gt;::encode):
(IPC::ArgumentCoder&lt;WebKit::AccessibilityPreferences&gt;::decode):
* Source/WebKit/Shared/AccessibilityPreferences.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::accessibilityPreferences):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::accessibilityPreferencesDidChange):

Canonical link: <a href="https://commits.webkit.org/251883@main">https://commits.webkit.org/251883@main</a>
</pre>

### CODE REVIEW

File: LayoutTests/TestExpectations [hunk 1]
```diff
@@ -5319,6 +5319,7 @@ fast/attachment/attachment-image-controls-basic.html [ Skip ]
 # This test requires a system font that has Bulgarian-specific shaping.
 fast/text/bulgarian-system-language-shaping.html [ ImageOnlyFailure ]
 
-# Temporarily disable accessibility bold tests while it's being reimplemented.
-webkit.org/b/237817 fast/text/accessibility-bold-system-font.html [ Pass Failure ]
-webkit.org/b/237817 fast/text/accessibility-bold-system-font-2.html [ Pass ImageOnlyFailure ]
+# Accessibility bold only exists on iOS.
+fast/text/accessibility-bold-system-font.html [ Pass Failure ]
```
Comment (smfr) [inline LayoutTests/TestExpectations]: `Pass Failure` is an odd way to express this. Just skip on all platforms that aren't iOS?

File: LayoutTests/fast/text/accessibility-bold-system-font-3-expected-mismatch.html [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html><!-- webkit-test-runner [ runSingly=true ] -->
+<html>
+<head>
+<script>
+if (window.internals) {
+    window.internals.invalidateFontCache();
```
Comment (smfr) [inline LayoutTests/fast/text/accessibility-bold-system-font-3-expected-mismatch.html]: Feels odd for a test to have to worry about caches. Why doesn't `window.internals.settings.setShouldMockBoldSystemFontForAccessibility` invalidate the cache?

File: LayoutTests/platform/ios/TestExpectations [hunk 3]
```diff
@@ -3659,3 +3659,7 @@ webkit.org/b/241048 imported/w3c/web-platform-tests/html/canvas/element/manual/w
 webkit.org/b/241205 fast/forms/textfield-outline.html [ Pass Failure ]
 
 fast/text/bulgarian-system-language-shaping.html [ Pass ]
+
+fast/text/accessibility-bold-system-font.html [ Pass ]
```
Comment (smfr) [inline LayoutTests/platform/ios/TestExpectations]: Might be a little nicer to move these tests into their own dir so you can just pass the whole dir.

File: Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp [hunk 4]
```diff
@@ -567,6 +566,28 @@ static void addAttributesForFontPalettes(CFMutableDictionaryRef attributes, cons
     }
 }
 
+static std::optional<bool>& overrideEnhanceTextLegibilityStorage()
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp]: Can you call this just `overrideEnhanceTextLegibility()` and call the one below `setOverrideEnhanceTextLegibility(bool)` ?

File: Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp [hunk 5]
```diff
@@ -567,6 +566,28 @@ static void addAttributesForFontPalettes(CFMutableDictionaryRef attributes, cons
...
+{
+    static NeverDestroyed<std::optional<bool>> overrideEnhanceTextLegibility;
+    return overrideEnhanceTextLegibility.get();
+}
+
+void overrideEnhanceTextLegibility(bool override)
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp]: I'm surprised the compiler let you use a keyword as a variable name.

**Review by smfr — APPROVED** [review]

File: LayoutTests/TestExpectations (hunk 1 above)
Comment (litherum) [reply LayoutTests/TestExpectations]: I thought `Pass Failure` was recommended, to make sure the test doesn't crash on other OSes?

File: LayoutTests/fast/text/accessibility-bold-system-font-3-expected-mismatch.html (hunk 2 above)
Comment (litherum) [reply LayoutTests/fast/text/accessibility-bold-system-font-3-expected-mismatch.html]: Good idea!

File: Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp (hunk 5 above)
Comment (litherum) [reply Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp]: C++ parsing is context-aware 😅

**Comment by litherum** [comment]
I want to make sure Windows builds before marking cq+

**Comment by litherum** [comment]
Running out of time, have to cq+ this now.

**Pull request merged**